<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <script type='text/javascript' src="./lib/shallowCompare.js"></script>

  <script type='text/javascript' src="./dist/React.js"></script>
   <script type='text/javascript' src="./lib/ReactTestUtils.js"></script>
 <!-- <script type='text/javascript' src="./test/react.js"></script>
  <script type='text/javascript' src="./test/react-dom.js"></script>-->
 <!-- <script type='text/javascript' src="./test/react.development.js"></script>
  <script type='text/javascript' src="./test/react-dom.development.js"></script>-->
  
  <!--<script type='text/javascript' src="./test/react-lite.js"></script>-->
  <script type='text/javascript' src="./lib/babel.js"></script>

</head>

<body>
  <div>开发者工具</div>
  <pre>
  </pre>
  <div id='example'></div>
  <script type='text/babel'>
   
   var container = document.getElementById("example");
   var div = container;
   // var PropTypes = React.PropTypes
   if (!window.ReactDOM) {
       window.ReactDOM = window.React;
   }
   var log = [];
   function logger(e) {
       log.push(e);
   }
   var PropTypes = React.PropTypes;
   var expect = function(a) {
       return {
           toBe: function(b) {
               console.log(a, "\nvs\n", b, a === b);
           }
       };
   };
   
   var ErrorBoundary = class extends React.Component {
       constructor(props) {
           super(props);
           this.state = { error: null };
           logger(`${this.props.logName} constructor`);
       }
       render() {
           if (this.state.error && !this.props.forceRetry) {
               logger(`${this.props.logName} render error`);
   
               return this.props.renderError(this.state.error, this.props);
           }
           logger(`${this.props.logName} render success`);
           return <div>{this.props.children}</div>;
       }
       componentDidCatch(error) {
           logger(`${this.props.logName} componentDidCatch`);
           this.setState({ error });
       }
       componentWillMount() {
           logger(`${this.props.logName} componentWillMount`);
       }
       componentDidMount() {
           logger(`${this.props.logName} componentDidMount`);
       }
       componentWillReceiveProps() {
           logger(`${this.props.logName} componentWillReceiveProps`);
       }
       componentWillUpdate() {
           logger(`${this.props.logName} componentWillUpdate`);
       }
       componentDidUpdate() {
           logger(`${this.props.logName} componentDidUpdate`);
       }
       componentWillUnmount() {
           logger(`${this.props.logName} componentWillUnmount`);
       }
   };
   ErrorBoundary.defaultProps = {
       logName: "ErrorBoundary",
       renderError(error, props) {
           return <div ref={props.errorMessageRef}>Caught an error: {error.message}.</div>;
       }
   };
   
   var BrokenRenderErrorBoundary = class extends React.Component {
       constructor(props) {
           super(props);
           this.state = { error: null };
           log.push("BrokenRenderErrorBoundary constructor");
       }
       render() {
           if (this.state.error) {
               log.push("BrokenRenderErrorBoundary render error [!]");
               throw new Error("Hello");
           }
           log.push("BrokenRenderErrorBoundary render success");
           return <div>{this.props.children}</div>;
       }
       componentWillMount() {
           log.push("BrokenRenderErrorBoundary componentWillMount");
       }
       componentDidMount() {
           log.push("BrokenRenderErrorBoundary componentDidMount");
       }
       componentWillUnmount() {
           log.push("BrokenRenderErrorBoundary componentWillUnmount");
       }
       componentDidCatch(error) {
           log.push("BrokenRenderErrorBoundary componentDidCatch");
           this.setState({ error });
       }
   };
   
   var BrokenRender = class extends React.Component {
       constructor(props) {
           super(props);
           log.push("BrokenRender constructor");
       }
       render() {
           log.push("BrokenRender render [!]");
           throw new Error("Hello");
       }
       componentWillMount() {
           log.push("BrokenRender componentWillMount");
       }
       componentDidMount() {
           log.push("BrokenRender componentDidMount");
       }
       componentWillReceiveProps() {
           log.push("BrokenRender componentWillReceiveProps");
       }
       componentWillUpdate() {
           log.push("BrokenRender componentWillUpdate");
       }
       componentDidUpdate() {
           log.push("BrokenRender componentDidUpdate");
       }
       componentWillUnmount() {
           log.push("BrokenRender componentWillUnmount");
       }
   };
   
   ReactDOM.render(
       <ErrorBoundary>
           <BrokenRenderErrorBoundary>
               <BrokenRender />
           </BrokenRenderErrorBoundary>
       </ErrorBoundary>,
       container
   );
   expect(container.firstChild.textContent).toBe("Caught an error: Hello.");
   expect(log.join("\n")).toBe(
       [
           "ErrorBoundary constructor",
           "ErrorBoundary componentWillMount",
           "ErrorBoundary render success",
           "BrokenRenderErrorBoundary constructor",
           "BrokenRenderErrorBoundary componentWillMount",
           "BrokenRenderErrorBoundary render success",
           "BrokenRender constructor",
           "BrokenRender componentWillMount",
           "BrokenRender render [!]",
           // The first error boundary catches the error
           // It adjusts state but throws displaying the message
           // Finish mounting with null children
           "BrokenRenderErrorBoundary componentDidMount",
           // Attempt to handle the error
           "BrokenRenderErrorBoundary componentDidCatch",
           "ErrorBoundary componentDidMount",
           "BrokenRenderErrorBoundary render error [!]",
           // Boundary fails with new error, propagate to next boundary
           "BrokenRenderErrorBoundary componentWillUnmount",
           // Attempt to handle the error again
           "ErrorBoundary componentDidCatch",
           "ErrorBoundary componentWillUpdate",
           "ErrorBoundary render error",
           "ErrorBoundary componentDidUpdate"
       ].join("\n")
   );
   
   log.length = 0;
   
 // ReactDOM.unmountComponentAtNode(container);
 // expect(log).toEqual(["ErrorBoundary componentWillUnmount"]);
      
    </script>

</body>

</html>